go/private/rules/cgo.bzl

```diff
@@ -118,29 +125,34 @@ def cgo_configure(go, srcs, cdeps, cppopts, copts, cxxopts, clinkopts):
-                    # use -L and -l instead of hard coding the path to the library in
-                    # the binary. This gives users more flexibility. The linker will add
-                    # rpaths later. We can't add them here because they are relative to
-                    # the binary location, and we don't know where that is.
-                    libname = lib.basename[len("lib"):lib.basename.rindex(".")]
-                    clinkopts.extend(["-L", lib.dirname, "-l", libname])
-                    inputs_direct.append(lib)
-                elif (lib.basename.startswith("lib") and
-                      has_versioned_shared_lib_extension(lib.basename)):
-                    # With a versioned shared library, we must use the full filename,
-                    # otherwise the library will not be found by the linker.
-                    libname = ":%s" % lib.basename
-                    clinkopts.extend(["-L", lib.dirname, "-l", libname])
-                    inputs_direct.append(lib)
-                else:
-                    lib_opts.append(lib.path)
+            for library_to_link in _cc_libraries_to_link(d):
+                for lib in _cc_libs(library_to_link):
+                    # If both static and dynamic variants are available, Bazel will only give
+                    # us the static variant. We'll get one file for each transitive dependency,
+                    # so the same file may appear more than once.
+                    if (lib.basename.startswith("lib") and
+                        has_simple_shared_lib_extension(lib.basename)):
+                        # If the loader would be able to find the library using rpaths,
+                        # use -L and -l instead of hard coding the path to the library in
+                        # the binary. This gives users more flexibility. The linker will add
+                        # rpaths later. We can't add them here because they are relative to
+                        # the binary location, and we don't know where that is.
+                        libname = lib.basename[len("lib"):lib.basename.rindex(".")]
+                        clinkopts.extend(["-L", lib.dirname, "-l", libname])
+                        inputs_direct.append(lib)
+                    elif (lib.basename.startswith("lib") and
+                          has_versioned_shared_lib_extension(lib.basename)):
+                        # With a versioned shared library, we must use the full filename,
+                        # otherwise the library will not be found by the linker.
+                        libname = ":%s" % lib.basename
+                        clinkopts.extend(["-L", lib.dirname, "-l", libname])
+                        inputs_direct.append(lib)
+                    else:
+                        if library_to_link.alwayslink:
+                            lib_opts.append(_WHOLE_ARCHIVE_FLAGS_BY_GOOS[go.mode.goos][0])
```

Comment:
That's a good point. I'd love to scrap `_WHOLE_ARCHIVE_FLAGS_BY_GOOS` and replace this code with something more robust, though I don't (yet) know how I would do that.

Reply:
Hmm, let's replace `_WHOLE_ARCHIVE_FLAGS_BY_GOOS` with a function:

```
def _library_args(go, library_to_link):
```

If `alwayslink` is false, it can just return a list with the file path. Otherwise, it can return a list with the file path surrounded by the appropriate flags. It can test `go.env["CC"]` to decide which flags to use.